README.md
# alfred-rs

[![alfred-rs](https://github.com/spamwax/alfred-workflow/actions/workflows/main.yml/badge.svg)](https://github.com/spamwax/alfred-workflow/actions/workflows/main.yml)
[![alfred-rs](https://github.com/spamwax/alfred-workflow/actions/workflows/lint.yml/badge.svg)](https://github.com/spamwax/alfred-workflow/actions/workflows/lint.yml)
[![crates.io/crates/alfred-rs](https://img.shields.io/crates/v/alfred-rs)](https://crates.io/crates/alfred-rs)
[![https://docs.rs/alfred-rs/badge.svg](https://docs.rs/alfred-rs/badge.svg)](https://docs.rs/crate/alfred-rs/)

Write [Workflows][] for [Alfred][alfred.app] app with ease!

This crate adds enhanced features and quality-of-life improvements to
[other alfred crate][alfred]'s basic functionality of generating items
for **Script Filter** types in Alfred.

Using this crate to create your workflows, you can
- Set up automatic update of workflow ([`updater`] module).
- Painlessly read/write data related to workflow (settings, cache data, ...) ([`data`] module).

## Documentation
For examples and complete documentation visit [API Documentation][].

[`updater`]: https://docs.rs/alfred-rs/latest/alfred_rs/updater/index.html
[`data`]: https://docs.rs/alfred-rs/latest/alfred_rs/data/index.html
[alfred]: https://crates.io/crates/alfred
[alfred.app]: http://www.alfredapp.com
[Workflows]: https://www.alfredapp.com/workflows/
[API Documentation]: http://docs.rs/alfred-rs

## Changelog

Change logs are now kept in a [separate document](./CHANGELOG.md).

## License

Licensed under either of
 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT) at your option.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.1] - 2022-07-10
### Changed
- Add a user-agent to reqwest::Client

## [0.7.0] - 2022-07-09
### Changed
- Replace failure crate with anyhow for error handling.
- Upgrade/Update reqwest, url crate
- Update some docs
- minimum rustc version is 1.56
- Bump version of dependencies in Cargo.toml
- Use github's actions for CI test.

## [0.6.0] - 2021-06-01
### Changed
- Clean-up tests.
- GithubReleaser now prefers urls ending with .alfredworkflow

## [0.5.1] - 2019-02-24
### Changed
- **Breaking Change**: `Data::load()` now takes one argument as the file name.
- Use Workflow's cache directory for storing temp. files
### Added
- Add a clear() method to Data struct.

## [0.4.3] - 2019-02-22
### Fixed
- Fix crate version for docs.rs

## [0.4.0] - 2018-07-04
### Added
- **Breaking changes**
- Methods that save data now accept `ref` instead of moving the value to be save.
### Fixed
- Checking for updates will now correctly make network calls after prior failures.


src/updater/tests.rs
use self::releaser::tests::setup_mock_server;
// #[cfg(not(feature = "ci"))]
use self::releaser::GithubReleaser;
use self::releaser::MOCK_RELEASER_REPO_NAME;
use super::*;
use std::ffi::OsStr;
use std::{thread, time};
use tempfile::Builder;
const VERSION_TEST: &str = "0.10.5";
const VERSION_TEST_NEW: &str = "0.11.1"; // should match what the mock server replies for new version.

#[test]
fn it_tests_settings_filename() {
    setup_workflow_env_vars(true);
    let updater_state_fn = Updater::<GithubReleaser>::build_data_fn().unwrap();
    assert_eq!(
        "workflow.B0AC54EC-601C-YouForgotTo___Name_Your_Own_Work_flow_-updater.json",
        updater_state_fn.file_name().unwrap().to_str().unwrap()
    );
}

#[test]
fn it_ignores_saved_version_after_an_upgrade_async() {
    // Make sure a freshly upgraded workflow does not use version info from saved state
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    {
        // Next check it reports a new version since mock server has a release for us
        let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        updater.set_interval(0);
        updater.init().expect("couldn't init worker");

        assert!(updater.update_ready().expect("couldn't check for update"));
        assert_eq!(VERSION_TEST, format!("{}", updater.current_version()));
    }

    // Mimic the upgrade process by bumping the version
    StdEnv::set_var("alfred_workflow_version", VERSION_TEST_NEW);
    {
        let updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        // Updater should pick up new version rather than using saved one
        assert_eq!(VERSION_TEST_NEW, format!("{}", updater.current_version()));
        updater.init().expect("couldn't init worker");
        // No more updates
        assert!(!updater.update_ready().expect("couldn't check for update"));
    }
}

#[test]
#[should_panic(
    expected = "HTTP status client error (400 Bad Request) for url (http://127.0.0.1:1234/releases/latest)"
)]
fn it_handles_server_error_async() {
    setup_workflow_env_vars(true);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    // Next check will be immediate
    updater.set_interval(0);
    updater.init().expect("couldn't init worker");
    let _m = setup_mock_server(400);
    // This should panic with a BadRequest (400) error.
    updater.update_ready().unwrap();
}

#[test]
fn it_caches_async_workers_payload() {
    setup_workflow_env_vars(true);

    first_check_after_installing_workflow();
    let _m = setup_mock_server(200);
    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    // Next check will be immediate
    updater.set_interval(0);
    updater.init().expect("couldn't init worker");
    assert!(updater.update_ready().expect("couldn't check for update"),);

    // Consequent calls to update_ready should cache the payload.
    let _m = setup_mock_server(400);
    assert!(updater.update_ready().expect("couldn't check for update"),);
    assert!(updater.update_ready().expect("couldn't check for update"),);
    assert!(updater.update_ready().expect("couldn't check for update"),);

    {
        let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        // Next check will be immediate
        updater.set_interval(0);
        updater.init().expect("couldn't init worker");
        assert!(updater.update_ready().is_err());
    }
}

#[test]
fn it_get_latest_info_from_releaser() {
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);

    {
        first_check_after_installing_workflow();
        // Blocking
        let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        // Next check will be immediate
        updater.set_interval(0);
        updater.init().expect("couldn't init worker");

        assert!(updater
            .update_ready()
            .expect("Blocking: couldn't check for update"));
    }
    {
        // Non-blocking
        let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        // Next check will be immediate
        updater.set_interval(0);
        // Start async worker
        updater.init().expect("couldn't init worker");
        let wait = time::Duration::from_millis(500);
        thread::sleep(wait);

        assert!(updater
            .try_update_ready()
            .expect("Non-blocking: couldn't check for update"));
    }
}

#[allow(clippy::cast_possible_wrap)]
#[test]
fn it_does_one_network_call_per_interval() {
    {
        setup_workflow_env_vars(true);
        let _m = setup_mock_server(200);
        let wait_time = 1;

        first_check_after_installing_workflow();

        let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
        // Next check will be immediate
        updater.set_interval(0);
        updater.init().expect("couldn't init worker");

        // Next update_ready will make a network call
        assert!(updater.update_ready().expect("couldn't check for update"));

        // Increase interval
        updater.set_interval(wait_time as i64);
        assert!(!updater.due_to_check());

        // make mock server return error. This way we can test that no network call was made
        // assuming Updater can read its cache file successfully
        let _m = setup_mock_server(503);
        let t = updater.update_ready();
        assert!(t.is_ok());
        // Make sure we still report update is ready
        assert!(t.unwrap());

        // Now we test that after interval has passed we will make a call
        let two_sec = time::Duration::from_secs(wait_time);
        thread::sleep(two_sec);
        {
            let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
            updater.set_interval(wait_time as i64);
            updater.init().expect("couldn't init worker");
            assert!(updater.due_to_check());

            // Since server is returning error, update_ready() should fail.
            let t = updater.update_ready();
            assert!(t.is_err());
        }
        {
            // Just making sure the next call will go through and return expected results.
            let _m = setup_mock_server(200);
            let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
            // Next check will be immediate
            updater.set_interval(0);
            updater.init().expect("couldn't init worker");
            assert!(updater.due_to_check());

            // Since server is ok, update_ready() should work.
            let t = updater.update_ready();
            assert!(t.is_ok());
            assert!(updater.update_ready().expect("couldn't check for update"));
        }
    }
}

#[test]
fn it_tests_download() {
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");

    // Next check will be immediate
    updater.set_interval(0);
    // Force current version to be really old.
    updater.set_version("0.0.1");
    updater.init().expect("couldn't init worker");

    // New update is available
    assert!(updater.update_ready().expect("couldn't check for update"));

    let download_fn = updater.download_latest();
    assert!(download_fn.is_ok());
    assert_eq!(
        "latest_release_YouForgotTo___Name_Your_Own_Work_flow_.alfredworkflow",
        download_fn
            .unwrap()
            .file_name()
            .expect("couldn't get download file name")
            .to_str()
            .expect("impossible?!")
    );
}

#[test]
#[should_panic(expected = "no release info")]
fn it_doesnt_download_without_release_info() {
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    updater.set_interval(864_000);

    assert!(!updater.due_to_check());
    updater.init().expect("couldn't init worker");

    assert!(updater.download_latest().is_err());

    // Since check time is due yet, following will just read cache without
    // getting any release info, hence the last line should panic
    assert!(!updater.update_ready().expect("couldn't check for update"));
    updater.download_latest().unwrap();
}

#[test]
fn it_downloads_after_getting_release_info() {
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    updater.set_interval(0);
    updater.init().expect("couldn't init worker");
    assert!(updater.download_latest().is_err());

    assert!(updater.update_ready().expect("couldn't check for update"));
    assert!(updater.download_latest().is_ok());
}

#[test]
fn it_tests_async_updates_1() {
    //
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    // Next check will be immediate
    updater.set_interval(0);
    updater.init().expect("couldn't init worker");
    assert!(updater.update_ready().expect("couldn't check for update"));
}

#[test]
fn it_tests_async_updates_2() {
    // This test will only spawn a thread once.
    // Second call will use a cache since it's not due to check.
    setup_workflow_env_vars(true);
    let _m = setup_mock_server(200);
    first_check_after_installing_workflow();

    let mut updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");

    // Next check will spawn a thread. There should be an update avail. from mock server.
    updater.set_interval(0);
    updater.init().expect("couldn't init worker");
    updater.update_ready().expect("chouldn't check for update");

    // make mock server return error. This way we can test that no network call was made
    // assuming Updater can read its cache file successfully
    let _m = setup_mock_server(503);
    // Increase interval
    updater.set_interval(86400);

    assert!(updater.update_ready().expect("couldn't check for update"));
}

pub(super) fn setup_workflow_env_vars(secure_temp_dir: bool) -> PathBuf {
    // Mimic Alfred's environment variables
    let path = if secure_temp_dir {
        Builder::new()
            .prefix("alfred_workflow_test")
            .rand_bytes(5)
            .tempdir()
            .unwrap()
            .into_path()
    } else {
        StdEnv::temp_dir()
    };
    {
        let v: &OsStr = path.as_ref();
        StdEnv::set_var("alfred_workflow_data", v);
        StdEnv::set_var("alfred_workflow_cache", v);
        StdEnv::set_var("alfred_workflow_uid", "workflow.B0AC54EC-601C");
        StdEnv::set_var(
            "alfred_workflow_name",
            "YouForgotTo/フ:Name好YouráOwnسWork}flowッ",
        );
        StdEnv::set_var("alfred_workflow_bundleid", "MY_BUNDLE_ID");
        StdEnv::set_var("alfred_workflow_version", VERSION_TEST);
    }
    path
}

fn first_check_after_installing_workflow() {
    // since the first check after workflow installation by user will return no update available,
    // we need to run it at the beginning of some tests
    let _m = setup_mock_server(200);

    let updater = Updater::gh(MOCK_RELEASER_REPO_NAME).expect("cannot build Updater");
    assert_eq!(VERSION_TEST, format!("{}", updater.current_version()));

    updater.init().expect("couldn't init worker");

    // First update_ready is always false.
    assert!(!updater.update_ready().expect("couldn't check for update"));
}


.github/workflows/run_tests.sh
#!/usr/bin/env bash

export RUST_TEST_NOCAPTURE=1
if [ -n "$CARGO_REGISTRY_TOKEN" ]; then
  echo "${CARGO_REGISTRY_TOKEN:1:3}"
else
  echo "didn't get CARGO_REGISTRY_TOKEN!"
fi
exit

cargo test --features updater --lib -- --test-threads=1
# cargo test --features updater,ci --lib -- --test-threads=1

# cargo test --features updater --lib -- --ignored --test-threads=1
# cargo test --features updater,ci --lib -- --ignored --test-threads=1

# doc tests
cargo test --features updater --doc
# cargo test --features updater,ci --doc



src/lib.rs
// Copyright (c) 2018 Hamid R. Ghadyani.
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Write [Workflows] for [Alfred][alfred.app] app with ease!
//!
//! This crate adds enhanced features and quality-of-life improvements to
//! [other alfred crate][alfred]'s basic functionality of creating **Script Filter** items.
//!
//! Using this crate to create your workflows, you can
//! - Set up automatic update of workflow ([`updater`] module).
//! - Painlessly read/write data related to workflow (settings, cache data, ...) ([`data`] module).
//!
//! [`updater`]: updater/index.html
//! [`data`]: data/index.html
//! [alfred]: https://crates.io/crates/alfred
//! [alfred.app]: http://www.alfredapp.com
//! [Workflows]: https://www.alfredapp.com/workflows/
//!

// TODO: check for "status" field of json returned by github to make sure it is fully uploaded
// before reporting that a release is available.
// TODO: Automatically update html_root_url's version when publishing to crates.io
// TODO: Use https://github.com/softprops/hubcaps for github API?

#![doc(html_root_url = "https://docs.rs/alfred-rs/0.7.1")]

extern crate alfred;
extern crate serde;
extern crate serde_json;

#[cfg(test)]
extern crate mockito;

#[macro_use]
extern crate log;
extern crate chrono;
extern crate env_logger;
extern crate semver;
#[macro_use]
extern crate serde_derive;
extern crate tempfile;
extern crate url;

use alfred::env;
use anyhow::Result;
use anyhow::{anyhow, bail};

pub mod data;
pub mod updater;

pub use self::data::Data;
pub use self::updater::Updater;


src/updater/releaser.rs
use super::{anyhow, reqwest, semver, serde_json, url, Result};
#[cfg(test)]
use mockito;
use semver::Version;
use std::cell::RefCell;
use url::Url;

#[cfg(not(test))]
const GITHUB_API_URL: &str = "https://api.github.com/repos/";
const GITHUB_LATEST_RELEASE_ENDPOINT: &str = "/releases/latest";

#[cfg(test)]
#[allow(deprecated)]
static MOCKITO_URL: &str = mockito::SERVER_URL;
#[cfg(test)]
pub const MOCK_RELEASER_REPO_NAME: &str = "MockZnVja29mZg==fd850fc2e63511e79f720023dfdf24ec";

/// An interface for checking with remote servers to identify the latest release for an
/// Alfred workflow.
///
/// This trait has been implemented for [`GithubReleaser`] to check for a newer version of a workflow
/// that's maintained on `github.com`
///
/// [`GithubReleaser`]: struct.GithubReleaser.html
pub trait Releaser: Clone {
    /// Typte that represents semantic compatible identifier of a release.
    type SemVersion: Into<Version>;

    /// Type that represents a url to the latest release resource.
    type DownloadLink: Into<Url>;

    /// Creates a new `Releaser` instance that is identified as `name`
    fn new<S: Into<String>>(name: S) -> Self;

    /// Performs necessary communications to obtain release info in form of
    /// `SemVersion` and `DownloadLink` types.
    ///
    /// Returned tuple consists of semantic version compatible identifier of the release and
    /// a download link/url that can be used to fetch the release.
    ///
    /// Implementors are strongly encouraged to get the meta-data about the latest release without
    /// performing a full download of the workflow.
    ///
    /// # Errors
    /// Method returns `Err(Error)` on file or network error.
    fn fetch_latest_release(&self) -> Result<(Self::SemVersion, Self::DownloadLink)>;

    /// Returns the latest release information that is available from server.
    ///
    /// # Errors
    /// Method returns `Err(Error)` on file or network error.
    fn latest_release(&self) -> Result<(Version, Url)> {
        let (v, url) = self.fetch_latest_release()?;
        Ok((v.into(), url.into()))
    }
}

/// Struct to handle checking and finding release files from `github.com`
///
/// This implementation of `Releaser` will favor files that end with `alfred3workflow`
/// over `alfredworkflow`. If there are multiple `alfred3workflow`s or `alfredworkflow`s, the first
/// one returned by `github.com` will be used.
///
/// See [`updater::gh()`] for how to use this.
///
/// [`updater::gh()`]: struct.Updater.html#method.gh
#[allow(clippy::module_name_repetitions)]
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GithubReleaser {
    repo: String,
    latest_release: RefCell<Option<ReleaseItem>>,
}

// Struct to store information about a single release point.
//
// Each release point may have multiple downloadable assets.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ReleaseItem {
    /// name of release that should hold a semver compatible identifier.
    pub tag_name: String,
    assets: Vec<ReleaseAsset>,
}

/// A single downloadable asset.
#[derive(Debug, Serialize, Deserialize, Clone)]
struct ReleaseAsset {
    url: String,
    name: String,
    state: String,
    browser_download_url: String,
}

impl GithubReleaser {
    fn latest_release_data(&self) -> Result<()> {
        debug!("starting latest_release_data");
        let client = reqwest::blocking::Client::builder()
            .user_agent(concat!(
                env!("CARGO_PKG_NAME"),
                "/",
                env!("CARGO_PKG_VERSION")
            ))
            .build()?;

        #[cfg(test)]
        let url = format!("{}{}", MOCKITO_URL, GITHUB_LATEST_RELEASE_ENDPOINT);

        #[cfg(not(test))]
        let url = format!(
            "{}{}{}",
            GITHUB_API_URL, self.repo, GITHUB_LATEST_RELEASE_ENDPOINT
        );
        debug!("  url is: {:?}", url);

        client
            .get(&url)
            .send()?
            .error_for_status()
            .map_err(Into::into)
            .and_then(|resp| {
                let mut latest: ReleaseItem = serde_json::from_reader(resp)?;
                if latest.tag_name.starts_with('v') {
                    latest.tag_name.remove(0);
                }
                debug!("  release item: {:?}", latest);
                *self.latest_release.borrow_mut() = Some(latest);
                Ok(())
            })
    }

    // This implementation of Releaser will favor urls that end with `alfredworkflow`
    // over `alfredworkflow`
    fn downloadable_url(&self) -> Result<Url> {
        debug!("starting download_url");
        self.latest_release
            .borrow()
            .as_ref()
            .ok_or_else(|| {
                anyhow!(
                "no release item available, did you first get version by calling latest_version?",
            )
            })
            .and_then(|r| {
                let urls = r
                    .assets
                    .iter()
                    .filter(|asset| {
                        asset.state == "uploaded"
                            && (asset.browser_download_url.ends_with("alfredworkflow")
                                || asset.browser_download_url.ends_with("alfred3workflow")
                                || asset.browser_download_url.ends_with("alfred4workflow"))
                    })
                    .map(|asset| &asset.browser_download_url)
                    .collect::<Vec<&String>>();
                debug!("  collected release urls: {:?}", urls);
                match urls.len() {
                    0 => Err(anyhow!("no usable download url")),
                    1 => Ok(Url::parse(urls[0])?),
                    _ => {
                        let url = urls.iter().find(|item| item.ends_with("alfredworkflow"));
                        let u = url.unwrap_or(&urls[0]);
                        Ok(Url::parse(u)?)
                    }
                }
            })
    }

    fn latest_version(&self) -> Result<Version> {
        debug!("starting latest_version");
        if self.latest_release.borrow().is_none() {
            self.latest_release_data()?;
        }

        let latest_version = self
            .latest_release
            .borrow()
            .as_ref()
            .map(|r| Version::parse(&r.tag_name).ok())
            .ok_or_else(|| anyhow!("Couldn't parse fetched version."))?
            .unwrap();
        debug!("  latest version: {:?}", latest_version);
        Ok(latest_version)
    }
}

impl Releaser for GithubReleaser {
    type SemVersion = Version;
    type DownloadLink = Url;

    fn new<S: Into<String>>(repo_name: S) -> GithubReleaser {
        GithubReleaser {
            repo: repo_name.into(),
            latest_release: RefCell::new(None),
        }
    }

    fn fetch_latest_release(&self) -> Result<(Version, Url)> {
        if self.latest_release.borrow().is_none() {
            self.latest_release_data()?;
        }
        let version = self.latest_version()?;
        let link = self.downloadable_url()?;
        Ok((version, link))
    }
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use mockito::{mock, Matcher, Mock};

    #[test]
    fn it_tests_releaser() {
        let _m = setup_mock_server(200);
        let releaser = GithubReleaser::new(MOCK_RELEASER_REPO_NAME);

        // Calling downloadable_url before checking for latest_version will return error
        assert!(releaser.downloadable_url().is_err());

        assert!(
            releaser
                .latest_version()
                .expect("couldn't do latest_version")
                > Version::new(0, 11, 0)
        );

        assert_eq!("http://127.0.0.1:1234/releases/download/v0.11.1/alfred-pinboard-rust-v0.11.1.alfredworkflow",
                   releaser.downloadable_url().unwrap().as_str());
    }

    pub fn setup_mock_server(status_code: usize) -> Mock {
        mock(
            "GET",
            Matcher::Regex(r"^/releases/(latest|download).*$".to_string()),
        )
        .with_status(status_code)
        .with_header("content-type", "application/json")
        .with_body(include_str!("../../tests/latest.json"))
        .create()
    }
}
